thunderfish/fakefish.py

```diff
@@ -1,123 +1,233 @@
-    :param tr_std: (float). std of the negative part of the pulse in sec.
-    :param tr_amplitude: (float). Factor for regulating the negative part of the pulse.
-    :param tr_time:
-
-    :return: (array). Data with pulses at the given frequency.
+    :param duration: (float). Duration of the generated data in seconds.
+    :param noise_std: (float). Standard deviation of additive Gaussian white noise.
+    :param jitter_cv: (float). Gaussian distributed jitter of pulse times as coefficient of variation of inter-pulse intervals.
+    :param peak_std: (float or list of floats). Standard deviation of Gaussian shaped peaks in seconds.
+    :param peak_amplitude: (float or list of floats). Amplitude of each peak (positive and negative).
+    :param peak_time: (float or list of floats). Position of each Gaussian peak in seconds.
+
+    :return data: (array). Generated data of a pulse-type fish.
     """
-    # Create a Gaussian Distribution; one for the peak and another for the trough
 
-    x = np.arange(-4.*pk_std, tr_time + 4.*tr_std, 1./samplerate)
+    # make sure peak properties are in a list:
+    if np.isscalar(peak_std):
+        peak_stds = [peak_std]
+        peak_amplitudes = [peak_amplitude]
+        peak_times = [peak_time]
+    else:
+        peak_stds = peak_std
+        peak_amplitudes = peak_amplitude
+        peak_times = peak_time
+
+    # time axis for single pulse:
+    min_time_inx = np.argmin(peak_times)
+    max_time_inx = np.argmax(peak_times)
+    x = np.arange(-4.*peak_stds[min_time_inx] + peak_times[min_time_inx],
+                  4.*peak_stds[max_time_inx] + peak_times[max_time_inx], 1.0/samplerate)
+    pulse_duration = x[-1] - x[0]
+    
+    # generate a single pulse:
+    pulse = np.zeros(len(x))
+    for time, ampl, std in zip(peak_times, peak_amplitudes, peak_stds):
+        pulse += ampl * np.exp(-0.5*((x-time)/std)**2) 
+
+    # paste the pulse onto the noise floor:
+    time = np.arange(0, duration, 1. / samplerate)
+    data = np.random.randn(len(time)) * noise_std
+    period = 1.0/frequency
+    jitter_std = period * jitter_cv
+    first_pulse = np.max(pulse_duration, 3.0*jitter_std)
+    pulse_times = np.arange(first_pulse, duration, period )
+    pulse_times += np.random.randn(len(pulse_times)) * jitter_std
+    pulse_indices = np.round(pulse_times * samplerate).astype(np.int)
+    for inx in pulse_indices[pulse_indices < len(data)-len(pulse)-1]:
+        data[inx:inx + len(pulse)] += pulse
 
-    pulse = np.exp(-0.5 * (x/pk_std)**2) * pk_amplitude  # Peak Gaussian
-    pulse -= np.exp(-0.5 * ((x-tr_time)/tr_std)**2) * tr_amplitude  # Trough Gaussian
+    return data
 
-    # Now we need to set the pulse into some baseline with noise.
-    time = np.arange(0, time_len, 1. / samplerate)
 
-    data = np.random.randn(len(time)) * noise_fac
+def generate_monophasic_pulses(frequency=100.0, samplerate=44100., duration=1.,
+                               noise_std=0.01, jitter_cv=0.1):
```

Comment:
so if I understood correctly, you just vary the noise_std and by that you get either a monophasic or a biphasic pulse? that's nice.